integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonIngestCommand.scala

```diff
@@ -0,0 +1,368 @@
+          // delete all stage files
+          lock.unlock()
+          lock = null
+          LOGGER.info(s"Segment $segmentId is in SUCCESS state, about to delete " +
+                      s"${stageFileNames.length} stage files")
+          val numThreads = Math.min(Math.max(stageFileNames.length, 1), 10)
+          val executorService = Executors.newFixedThreadPool(numThreads)
+          stageFileNames.map { fileName =>
+            executorService.submit(new Runnable {
+              override def run(): Unit = {
+                FileFactory.getCarbonFile(
+                  CarbonTablePath.getStageDir(table.getTablePath) +
+                  CarbonCommonConstants.FILE_SEPARATOR + fileName
+                ).delete()
+              }
+            })
+          }.map { future =>
+            future.get()
+          }
+        case SegmentStatus.INSERT_IN_PROGRESS =>
+          // delete entry in table status and load again
+          LOGGER.info(s"Segment $segmentId is in INSERT_IN_PROGRESS state, about to delete the " +
+                      s"segment entry and load again")
+          val segmentToWrite = segments.filterNot(_.getLoadName.equals(segmentId))
+          SegmentStatusManager.writeLoadDetailsIntoFile(
+            CarbonTablePath.getTableStatusFilePath(table.getTablePath),
+            segmentToWrite)
+        case other =>
+          throw new RuntimeException(s"Segment $segmentId is in unexpected state: $other")
+      }
+    } finally {
+      if (lock != null) {
+        lock.unlock()
+      }
+    }
+    LOGGER.info(s"Finish recovery, delete snapshot file: $snapshotFilePath")
+    FileFactory.getCarbonFile(snapshotFilePath).delete()
+  }
+
+  /**
+   * Start global sort loading
+   */
+  private def startLoading(
+      spark: SparkSession,
+      table: CarbonTable,
+      loadModel: CarbonLoadModel,
+      stageInput: Seq[StageInput]
+  ): Unit = {
+    val splits = stageInput.flatMap(_.createSplits().asScala)
+    LOGGER.info(s"start to load ${splits.size} files into " +
+                s"${table.getDatabaseName}.${table.getTableName}")
+    val start = System.currentTimeMillis()
+    val dataFrame = DataLoadProcessBuilderOnSpark.createInputDataFrame(spark, table, splits)
+    DataLoadProcessBuilderOnSpark.loadDataUsingGlobalSort(
+      spark,
+      Option(dataFrame),
+      loadModel,
+      SparkSQLUtil.sessionState(spark).newHadoopConf()
+    ).map { row =>
+        (row._1, FailureCauses.NONE == row._2._2.failureCauses)
+      }
+
+    LOGGER.info(s"finish data loading, time taken ${System.currentTimeMillis() - start}ms")
+  }
+
+  /**
+   * Read stage files and return input files
+   */
+  private def collectStageInputs(
+      executorService: ExecutorService,
+      stageFiles: Array[(CarbonFile, CarbonFile)]
+  ): Seq[StageInput] = {
+    val startTime = System.currentTimeMillis()
+    val output = Collections.synchronizedList(new util.ArrayList[StageInput]())
+    val gson = new Gson()
```

Comment:
Gson is thead-safe object, so we can consider to reuse it.

Reply:
I will do it in another PR, to modify all gson usage